src/linalg/nonbacktracking.jl

```diff
@@ -26,17 +26,22 @@ function non_backtracking_matrix(g::AbstractGraph)
         end
     end
 
-    B = zeros(Float64, m, m)
-
+    nz = if is_directed(g)
+        sum(indegree(g) .* outdegree(g))
+    else
+        sum(degree(g) .* (degree(g) .- 1))
+    end
+    rowidx = sizehint!(Vector{Int}(), nz)
+    colidx = sizehint!(Vector{Int}(), nz)
     for (e, u) in edgeidmap
         i, j = src(e), dst(e)
-        for k in inneighbors(g, i)
-            k == j && continue
-            v = edgeidmap[Edge(k, i)]
-            B[v, u] = 1
-        end
+
+        rows = [edgeidmap[Edge(k, i)] for k in inneighbors(g, i) if k != j]
```

Comment:
iterating over neighbors like you did before would be just as efficient, and lead to less allocations

Reply:
Is this what you have in mind for the code?

```julia
rows = sizehint!(Vector{Int}(), indegree(g, i))
for k in inneighbors(g, i)
    k == j && continue
    push!(rows, edgeidmap[Edge(k, i)])
end
append!(rowidx, rows)
```

Reply:
why do you need to store the rows before iterating on them? just do it within a loop over neighbors

Reply:
I modified this part. If the code does not reflect your intentions, I would appreciate it if you could correct it.

Reply:
yeah that's good